tensorflow_quantum/core/qsim/state_space_test.cc

```diff
@@ -18,15 +18,380 @@ limitations under the License.
+                               0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
+                               0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0,
+                               0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0};
+  state->SetStateZero();
+  state->ApplyGate2(0, 3, matrix_hi);
+  ASSERT_EQ(state->GetAmpl(0), std::complex<float>(1 / std::sqrt(2), 0.0));
+  ASSERT_EQ(state->GetAmpl(8), std::complex<float>(1 / std::sqrt(2), 0.0));
+  state->ApplyGate2(0, 3, matrix_cnot);
+  ASSERT_EQ(state->GetAmpl(0), std::complex<float>(1 / std::sqrt(2), 0.0));
+  ASSERT_EQ(state->GetAmpl(9), std::complex<float>(1 / std::sqrt(2), 0.0));
+}
+
+TEST(StateSpaceTest, Update) {
+  const std::array<float, 8> matrix_h = {
+      1.0 / std::sqrt(2), 0.0, 1.0 / std::sqrt(2),  0.0,
+      1.0 / std::sqrt(2), 0.0, -1.0 / std::sqrt(2), 0.0};
+  Gate gate_small(0, 0, matrix_h);
+  std::vector<Gate> gates_small;
+  gates_small.push_back(gate_small);
+  const Circuit circuit_small(1, gates_small);
+  auto state_small = std::unique_ptr<StateSpace>(GetStateSpace(1, 1));
+  state_small->CreateState();
+  state_small->SetStateZero();
+  ASSERT_EQ(state_small->Update(circuit_small), tensorflow::Status::OK());
+  EXPECT_NEAR(state_small->GetAmpl(0).real(), 1.0 / std::sqrt(2), 1E-5);
+  EXPECT_NEAR(state_small->GetAmpl(0).imag(), 0.0, 1E-5);
+  EXPECT_NEAR(state_small->GetAmpl(1).real(), 1.0 / std::sqrt(2), 1E-5);
+  EXPECT_NEAR(state_small->GetAmpl(1).imag(), 0.0, 1E-5);
+
+  // TODO(zaqqwerty): Remove identity gates once fuser is updated (#171)
+  const uint64_t num_qubits(7);
+  const uint64_t q0(2);
+  const uint64_t q1(5);
+  Gate gate_0(0, q0, matrix_h);
+  Gate gate_1(1, q1, matrix_h);
+  const std::array<float, 32> matrix_i = {
+      1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0,
+      0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0,
+      0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0};
+  Gate gate_i(2, q0, q1, matrix_i);
+  std::vector<Gate> gates;
+  gates.push_back(gate_0);
+  gates.push_back(gate_1);
+  gates.push_back(gate_i);
+  const Circuit circuit(num_qubits, gates);
+  auto state = std::unique_ptr<StateSpace>(GetStateSpace(num_qubits, 1));
+  state->CreateState();
+  state->SetStateZero();
+  ASSERT_EQ(state->Update(circuit), tensorflow::Status::OK());
+  for (uint64_t i = 0; i < state_small->GetDimension(); i++) {
+    if (i == 0 || i == 1 << q0 || i == 1 << q1 || i == (1 << q0) + (1 << q1)) {
+      EXPECT_NEAR(state->GetAmpl(i).real(), 0.5, 1E-5);
+      EXPECT_NEAR(state->GetAmpl(i).imag(), 0.0, 1E-5);
+    } else {
+      EXPECT_NEAR(state->GetAmpl(i).real(), 0.0, 1E-5);
+      EXPECT_NEAR(state->GetAmpl(i).imag(), 0.0, 1E-5);
+    }
+  }
+}
+
+TEST(StateSpaceTest, ComputeExpectation) {
+  const uint64_t num_qubits(9);
+  const uint64_t q0(3);
+  const uint64_t q1(7);
+  auto state = std::unique_ptr<StateSpace>(GetStateSpace(num_qubits, 1));
+  auto scratch = GetStateSpace(num_qubits, 1);
```

Comment:
Why not put this in a smart pointer too ?

Reply:
Changed to smart pointer and used `get` to hand `ComputeExpectation` a raw pointer